### ISSUE TITLE
FSDP with gradient checkpointing lead to redundant allgathers

### ISSUE THREAD

**Issue by hanzhi713** [pytorch/pytorch#105635]
### 🐛 Describe the bug

While training huggingface Llama 13B with FSDP with full shard and gradient checkpointing enabled, I observed that the backward pass has two allgathers per layer. Compared to non checkpointed training, this additional allgather also affect the overlap between the reduce scatter and gradient computation.  I think ideally only one allgather is needed to gather the weights per FSDP module. 

Trace: (browm is reduce scatter, blue is allgather)
![allgather](https://github.com/pytorch/pytorch/assets/21034047/96d6527f-94cb-4c94-9256-42127eb663c0)

```python3
        model = FSDP(model,
                    sharding_strategy=ShardingStrategy.FULL_SHARD,
                    sync_module_states=True,
                    mixed_precision=mixed_precision,
                    auto_wrap_policy=functools.partial(transformer_auto_wrap_policy, transformer_layer_cls={LlamaDecoderLayer}),
                    limit_all_gathers=True,
                    device_id=dev,
                    param_init_fn=param_init_fn,
                    )
```

### Versions

I'm using pytorch nightly: `2.1.0.dev20230709+cu121`

**Comment by awgu** [comment]
How are you applying checkpointing?

**Comment by hanzhi713** [comment]
Huggingface's llama has built-in checkpointing. I simply turned it on via a flag. 

https://github.com/huggingface/transformers/blob/9ef5256dfb28f5419649af8ea94e82573a6b2e77/src/transformers/models/llama/modeling_llama.py#L690-L696

**Comment by hanzhi713** [comment]
Here's the trace of a single worker. Let me know if you need all workers. 

[llama.tar.gz](https://github.com/pytorch/pytorch/files/12113672/llama.tar.gz)

**Comment by awgu** [comment]
Thanks!

I will find some time to look at the traces soon, but in the short term, could you try:
1. Do not enable checkpointing via the built-in flag
2. Instead use the following:
```
from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
    checkpoint_wrapper,
    CheckpointImpl,
    apply_activation_checkpointing_wrapper,
)

auto_wrap_policy = functools.partial(transformer_auto_wrap_policy, transformer_layer_cls={LlamaDecoderLayer})
model = FSDP(model, ..., auto_wrap_policy)
non_reentrant_wrapper = functools.partial(checkpoint_wrapper, checkpoint_impl=CheckpointImpl.NO_REENTRANT)
apply_activation_checkpointing_wrapper(
    model, checkpoint_wrapper_fn=non_reentrant_wrapper, auto_wrap_policy=auto_wrap_policy
)
```

**Comment by hanzhi713** [comment]
With your new code, the problem seems to disappear. Hmm interesting

![image](https://github.com/pytorch/pytorch/assets/21034047/66fb6a56-7c45-4a35-be97-e4cef610e807)

**Comment by awgu** [comment]
There is some complexity in how FSDP and activation checkpointing interoperate. The FSDP changes in https://github.com/pytorch/pytorch/pull/105090 might help your original use case work, but I am not sure.

We have not landed those changes yet though.
